### PR TITLE
[libc++] Make sure LWG2070 is implemented as a DR

### LINKED ISSUE #54365: Bug in either C++17 (maybe 20?) spec, or llvm implementation, shared_ptr
Hey there,

I'm not sure where to post this, I feel like I should be posting to some spec committee.  Not sure.

In the Xcode 12.5.1 std memory, there is the snippet:
```

template <class _Tp, class _Alloc>
void
__shared_ptr_emplace<_Tp, _Alloc>::__on_zero_shared() _NOEXCEPT
{
    __data_.second().~_Tp();
}

```

What this snippet does, is when a control block of a shared_ptr detects that there are no more references, it destructs/destroys the object.

Right now the code snippet calls the destructor explicitly, which leaves out the exciting things the allocator might do before destructing the object.

I believe this code snippet should be something like this (writing off the top of my head):

```
using value_allocator = _Alloc::template rebind<the_value_type>::other;
value_allocator allocator;
allocator.destroy(&__data_.second());

```

Or using allocator_traits, or what not.  But in the end, not calling the destructor, but rather using the allocator to call the destructor.

---

If this were the case I could write an allocator which might do something like:

```
void destroy(T *p)
{
  object_is_about_to_be_destroyed_template_function(p);
  other_allocator.destroy(p);
}

```
---

Should I take this issue somewhere else?  Does anyone know who I might talk to about this.

---

There is, of course, a work around, but it isn't pretty, and uses classes which I probably should be using.

### PR DESCRIPTION
**Pull request by ldionne**
When we implemented C++20's P0674R1, we didn't enable the part of P0674R1 that was resolving LWG2070 as a DR. This patch fixes that and makes sure that we consistently go through the allocator when constructing and destroying the underlying object in std::allocate_shared.

Fixes #54365.

### CODE REVIEW

**Review by EricWF — CHANGES_REQUESTED** [review]
New tests please. Presumable this is observable, so we should have additional tests.

**Comment by ldionne** [comment]
> New tests please. Presumable this is observable, so we should have additional tests.

See the changes in `UNSUPPORTED` in `allocate_shared.lwg2070.pass.cpp` -- that's all that should be required.

File: libcxx/include/__memory/shared_ptr.h
Comment (EricWF) [inline libcxx/include/__memory/shared_ptr.h:326]: Can't we just use the other branch of code here? Why keep the `#ifdef` around. 

Especially because this code is in a vtable, and so it's an ABI issue. We can't have `#ifdef`s in code like this  Especially ifdefs that change on language dialect.

**Review by EricWF — CHANGES_REQUESTED** [review]

File: libcxx/include/__memory/shared_ptr.h
Comment (ldionne) [reply libcxx/include/__memory/shared_ptr.h:326]: You make a great point about this `#ifdef` being in a virtual function, which could lead to ODR violations if two TUs both contain a definition of the function but are compiled with different standard modes. We can't drop the `#if` entirely because `if constexpr` doesn't work before C++17, and this code needs to work all the way back to C++03.

That being said, any ODR violation would necessarily be benign, so I don't think this really has been a problem. Indeed, `__for_overwrite_tag` is only passed when `allocate_shared_for_overwrite` is used, and that function is only provided in C++20 mode. Hence, in a TU compiled with C++17 and earlier, this function would always be as follows (if we used the same definition in all standard modes):

```
if constexpr (false) {
    __get_elem()->~_Tp();
} else {
    using _TpAlloc = typename __allocator_traits_rebind<_Alloc, _Tp>::type;
    _TpAlloc __tmp(*__get_alloc());
    allocator_traits<_TpAlloc>::destroy(__tmp, __get_elem());
}
```

So whether the linker decides to select that function or the function defined as

```
using _TpAlloc = typename __allocator_traits_rebind<_Alloc, _Tp>::type;
_TpAlloc __tmp(*__get_alloc());
allocator_traits<_TpAlloc>::destroy(__tmp, __get_elem());
```

is irrelevant. That being said, I think it's easy to fix this problem so I'll do it in this patch. And this certainly was a good catch!

**Comment by ldionne** [comment]
The CI has passed and I think I've addressed the original comments on this, so I'll go ahead and merge this.

We did discuss the possibility of using `if constexpr` unconditionally instead of SFINAE since that is supported before C++17 as an extension. However, I tried that just now and this still causes problems because our test suite warns about using extensions in our headers. To unblock this, I'll go ahead and merge but I have a patch ready to go that NFC refactors this code back to `if constexpr`.

**Comment by dwblaikie** [comment]
Looks like this commit was made without approval? Were all the concerns addressed/did someone approve this through another medium?

**Comment by ldionne** [comment]
> Looks like this commit was made without approval? Were all the concerns addressed/did someone approve this through another medium?

We discussed it on Discord and like I recorded above, the original concerns were addressed. I committed it with the usual understanding that if there is something that needs to be addressed, it can be addressed post-commit like is quite usual in LLVM. Do you have any concrete concerns?

**Comment by dwblaikie** [comment]
> > Looks like this commit was made without approval? Were all the concerns addressed/did someone approve this through another medium?
> 
> We discussed it on Discord and like I recorded above, the original concerns were addressed. 

Ah, yeah - sorry, I saw your comments after I posted my reply (I'd just skimmed looking for approval, or informal approval in the form of a comment where someone might've forgot to pick the "approve" button) - didn't mean to be obtuse.

> I committed it with the usual understanding that if there is something that needs to be addressed, it can be addressed post-commit like is quite usual in LLVM. Do you have any concrete concerns?

Nah, just process concerns. It'd be good to get someone to punch the approve button if that's the consensus from Discord, etc - just for appearances/norms - I worry about newer contributors getting the wrong impression about making value calls about whether the review is "enough" before committing & hope we all wait for formal approval once something's sent for formal review.

**Comment by ldionne** [comment]
> > > Looks like this commit was made without approval? Were all the concerns addressed/did someone approve this 
> Nah, just process concerns. It'd be good to get someone to punch the approve button if that's the consensus from Discord, etc - just for appearances/norms - I worry about newer contributors getting the wrong impression about making value calls about whether the review is "enough" before committing & hope we all wait for formal approval once something's sent for formal review.

Yeah, I understand. I would argue that with Github PRs, most contributors shouldn't even need commit access, and then it would be the various maintainers' job to merge PRs. That would make this unambiguous. However I don't expect that this opinion is going to be very popular.

**Comment by dwblaikie** [comment]
> > > > Looks like this commit was made without approval? Were all the concerns addressed/did someone approve this
> > > > Nah, just process concerns. It'd be good to get someone to punch the approve button if that's the consensus from Discord, etc - just for appearances/norms - I worry about newer contributors getting the wrong impression about making value calls about whether the review is "enough" before committing & hope we all wait for formal approval once something's sent for formal review.
> 
> Yeah, I understand. I would argue that with Github PRs, most contributors shouldn't even need commit access, and then it would be the various maintainers' job to merge PRs. That would make this unambiguous. However I don't expect that this opinion is going to be very popular.

Yeah, I think we'll probably get there eventually - the days of review-after-commit being scalable/feasible/actually providing any checks and balances are probably well behind us & our process has some catching up to do with the modern scale of the project.

**Comment by ldionne** [comment]
We do have a review bandwidth problem though. A ton of changes end up never reviewed (and never committed). It's not great, but when there's not enough review bandwidth or interest, the solution is either to drop the change or to commit it without proper review. I think that's a problem we need to acknowledge if we want to make any improvements to our process.

**Comment by dwblaikie** [comment]
> We do have a review bandwidth problem though. A ton of changes end up never reviewed (and never committed). It's not great, but when there's not enough review bandwidth or interest, the solution is either to drop the change or to commit it without proper review. I think that's a problem we need to acknowledge if we want to make any improvements to our process.

Yep. For sure.

**Pull request merged**
